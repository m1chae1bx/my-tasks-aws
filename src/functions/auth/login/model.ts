import { JSONSchemaType, ajv } from "@libs/generic/util";

export interface LoginRequest {
  body: {
    id: string;
    password: string;
  };
}

export const loginRequestSchema: JSONSchemaType<LoginRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    body: {
      type: "object",
      properties: {
        id: {
          type: "string",
          maxLength: 100,
        },
        password: {
          type: "string",
        },
      },
      required: ["id", "password"],
    },
  },
  required: ["body"],
};

export const validateRequest = ajv.compile(loginRequestSchema);
