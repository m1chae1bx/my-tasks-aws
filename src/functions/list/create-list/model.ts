import { JSONSchemaType, ajv } from "/opt/nodejs/util";

export interface CreateListRequest {
  pathParameters: {
    userId: string;
  };
  body: {
    name: string;
    isDefault?: boolean;
  };
}

export const createListRequestSchema: JSONSchemaType<CreateListRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    pathParameters: {
      type: "object",
      properties: {
        userId: {
          type: "string",
        },
      },
      required: ["userId"],
    },
    body: {
      type: "object",
      properties: {
        name: {
          type: "string",
          maxLength: 100,
        },
        isDefault: {
          type: "boolean",
          nullable: true,
        },
      },
      required: ["name"],
    },
  },
  required: ["pathParameters", "body"],
};

export const validateRequest = ajv.compile(createListRequestSchema);
