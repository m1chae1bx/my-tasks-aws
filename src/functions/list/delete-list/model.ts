import { JSONSchemaType, ajv } from "/opt/nodejs/util";

export interface DeleteListRequest {
  pathParameters: {
    userId: string;
    listId: string;
  };
}

export const deleteListRequestSchema: JSONSchemaType<DeleteListRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    pathParameters: {
      type: "object",
      properties: {
        userId: {
          type: "string",
        },
        listId: {
          type: "string",
        },
      },
      required: ["userId", "listId"],
    },
  },
  required: ["pathParameters"],
};

export const validateRequest = ajv.compile(deleteListRequestSchema);
