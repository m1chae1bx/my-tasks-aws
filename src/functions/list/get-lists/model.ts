import { JSONSchemaType, ajv } from "/opt/nodejs/util";

export interface GetListsRequest {
  pathParameters: {
    userId: string;
  };
}

export const getListsRequestSchema: JSONSchemaType<GetListsRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    pathParameters: {
      type: "object",
      properties: {
        userId: {
          type: "string",
        },
      },
      required: ["userId"],
    },
  },
  required: ["pathParameters"],
};

export const validateRequest = ajv.compile(getListsRequestSchema);
