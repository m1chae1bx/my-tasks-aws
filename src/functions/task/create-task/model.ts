import { JSONSchemaType, ajv } from "@libs/generic/util";

export interface CreateTaskRequest {
  pathParameters: {
    listId: string;
  };
  body: {
    name: string;
    desc?: string;
    dueDate?: string;
  };
}

export const createTaskRequestSchema: JSONSchemaType<CreateTaskRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    pathParameters: {
      type: "object",
      properties: {
        listId: {
          type: "string",
        },
      },
      required: ["listId"],
    },
    body: {
      type: "object",
      properties: {
        name: {
          type: "string",
          maxLength: 200,
        },
        desc: {
          type: "string",
          nullable: true,
          maxLength: 400,
        },
        dueDate: {
          type: "string",
          nullable: true,
          format: "date-time",
        },
      },
      required: ["name"],
    },
  },
  required: ["pathParameters", "body"],
};

export const validateRequest = ajv.compile(createTaskRequestSchema);
