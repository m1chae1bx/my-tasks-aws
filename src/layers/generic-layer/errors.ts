export interface CustomError extends Error {
  errorCode: string;
}

export class UsernameUnavailableError implements CustomError {
  name = "UsernameUnavailableError";
  message = "Username is already taken";
  errorCode = "usernameUnavailable";
}

export class EmailUnavailableError implements CustomError {
  name = "EmailUnavailableError";
  message = "Email address is already registered";
  errorCode = "emailUnavailable";
}
