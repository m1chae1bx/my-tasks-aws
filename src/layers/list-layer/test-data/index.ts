import { ListDetails } from "../list.model";

export const testList: ListDetails = {
  id: "test-list-id",
  name: "Test list 1",
  userId: "test-user-id",
  isDefault: true,
};

export const testLists: ListDetails[] = [
  {
    id: "test-list-id-1",
    name: "Test list 1",
    userId: "test-user-id",
    isDefault: true,
  },
  {
    id: "test-list-id-2",
    name: "Test list 2",
    userId: "test-user-id",
    isDefault: false,
  },
];
