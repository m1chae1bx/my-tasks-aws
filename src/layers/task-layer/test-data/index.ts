import { DueDate, GetTasksQuery, TaskDetails } from "../task.model";

export const testTask: TaskDetails = {
  id: "test-task-id",
  listId: "test-list-id",
  name: "Test task",
  isCompleted: false,
  dueDate: new Date("2022-01-01"),
  desc: "Test task description",
};

export const completedTask: TaskDetails = {
  ...testTask,
  isCompleted: true,
};

export const testTasks: Partial<TaskDetails>[] = [
  {
    id: "test-task-id-1",
    name: "Test task 1",
    desc: "Test task description",
    isCompleted: false,
    dueDate: new Date("2020-01-01T00:00:00.000Z"),
    listId: "test-list-id",
  },
  {
    id: "test-task-id-2",
    name: "Test task 2",
    desc: "Test task description",
    isCompleted: true,
    dueDate: new Date("2020-01-01T00:00:00.000Z"),
    listId: "test-list-id",
  },
];

export const getTasksQuery: GetTasksQuery = {
  name: "Test task",
  dueDate: DueDate.TODAY,
  today: "2020-01-01",
  includeCompleted: "true",
};
