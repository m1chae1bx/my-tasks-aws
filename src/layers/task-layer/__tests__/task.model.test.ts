import { Task } from "../task.model";
import { testTask } from "../test-data";
import * as TaskDao from "../task.dao";
import { ErrorCode, RequiredPropertyMissingError } from "/opt/nodejs/errors";

jest.mock("../task.dao", () => ({
  ...jest.requireActual("../task.dao"),
  create: jest.fn(),
  update: jest.fn(),
  getAll: jest.fn(),
  deleteTask: jest.fn(),
}));

jest.useFakeTimers().setSystemTime(new Date("2022-01-01"));

describe("Task", () => {
  describe("constructor", () => {
    it("should initialize a task", () => {
      const { id, listId, name, isCompleted, dueDate, desc } = testTask;
      const task = new Task(
        listId,
        name,
        isCompleted,
        dueDate?.toISOString(),
        desc,
        id
      );
      expect(task.id).toBe(id);
      expect(task.listId).toBe(listId);
      expect(task.name).toBe(name);
      expect(task.isCompleted).toBe(isCompleted);
      expect(task.dueDate?.toISOString()).toBe(dueDate?.toISOString());
      expect(task.desc).toBe(desc);
    });
  });

  describe("object methods", () => {
    let task: Task;

    beforeEach(() => {
      const { id, listId, name, isCompleted, dueDate, desc } = testTask;
      task = new Task(
        listId,
        name,
        isCompleted,
        dueDate?.toISOString(),
        desc,
        id
      );
    });

    describe("save", () => {
      it("should return the task ID", async () => {
        (TaskDao.create as jest.Mock)
          .mockClear()
          .mockResolvedValueOnce("test-task-id");
        const result = await task.save();
        expect(result).toBe("test-task-id");
      });
    });

    describe("update - happy path", () => {
      it("should call TaskDao.update", async () => {
        (TaskDao.update as jest.Mock).mockClear();
        await task.update();
        expect((TaskDao.update as jest.Mock).mock.calls).toMatchSnapshot();
      });
    });

    describe("update - missing ID", () => {
      it("should throw an error", async () => {
        task.id = undefined;
        try {
          await task.update();
        } catch (error) {
          expect((error as RequiredPropertyMissingError).errorCode).toBe(
            ErrorCode.REQUIRED_PROPERTY_MISSING
          );
        }
      });
    });
  });

  describe("static getAll", () => {
    it("should return a list of tasks", async () => {
      (TaskDao.getAll as jest.Mock)
        .mockClear()
        .mockResolvedValueOnce([testTask]);
      const result = await Task.getAll("test-list-id", {});
      expect(result[0]).toEqual(testTask);
    });
  });

  describe("static delete", () => {
    it("should call TaskDao.delete", async () => {
      (TaskDao.deleteTask as jest.Mock).mockClear().mockResolvedValueOnce(null);
      await Task.delete("test-task-id", "test-list-id");
      expect((TaskDao.deleteTask as jest.Mock).mock.calls).toMatchSnapshot();
    });
  });
});
