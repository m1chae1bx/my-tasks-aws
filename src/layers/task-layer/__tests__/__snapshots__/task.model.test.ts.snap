// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task object methods update - happy path should call TaskDao.update 1`] = `
Array [
  Array [
    Object {
      "desc": "Test task description",
      "dueDate": 2022-01-01T00:00:00.000Z,
      "id": "test-task-id",
      "isCompleted": false,
      "listId": "test-list-id",
      "name": "Test task",
    },
  ],
]
`;

exports[`Task static delete should call TaskDao.delete 1`] = `
Array [
  Array [
    "test-task-id",
    "test-list-id",
  ],
]
`;
